### PR TITLE
feat: Handle `no_predicates` attribute

### PR DESCRIPTION
**Pull request by vezenovm**
# Description

## Problem\*

Resolves https://github.com/noir-lang/noir/issues/4911 and https://github.com/noir-lang/noir/issues/4688

## Summary\*

~~We recently included the `#[inline(never)]` attribute to enable developers to optimize codegen.~~ This has now been switched to the name `no_predicates`. The main use-case in mind is for circuits in issue #4688 where inlining a function with heavy array operations when dependent upon witnesses is not always ideal. Specifically when the function being inlined does not need to rely on the predicate for correctness.  

Originally I had in mind to delay inlining all the way to after ACIR gen and inline the ACIR artifacts. However, this feels overly complex now as we have all the infrastructure to inline functions as we wish during SSA, we could just need to delay the inlining of certain functions to happen after flattening. This PR does exactly what was just mentioned.

For example, the new test `no_predicates_numeric_generic_poseidon` gave these results when `poseidon_hash` was not marked with `#[no_predicates]`:
ExistingInlineStrategy

While when `poseidon_hash` was marked with `#[no_predicates]`:
InlineNeverBench

## Documentation\*

Check one:
- [ ] No documentation needed.
- [ ] Documentation included in this PR.
- [X] **[For Experimental Features]** Documentation to be submitted in a separate PR.

# PR Checklist\*

- [X] I have tested the changes locally.
- [X] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

**Comment by vezenovm** [comment]
Look to be getting some panics still so can hold off on a review

**Comment by vezenovm** [comment]
I was thinking to perhaps change the name of the attribute to something more specific to the Noir codegen. Like `#[inline(post_if_flatten)]` or `#[inline(post_flatten)]`. It is a bit more verbose but it gets across better what is actually happening and how it provides an optimization when writing circuits. Documentation should also help with any knowledge gap (which will come in a follow-up).

File: compiler/noirc_evaluator/src/ssa.rs [hunk 1]
```diff
@@ -61,6 +61,9 @@ pub(crate) fn optimize_into_acir(
         .run_pass(Ssa::remove_bit_shifts, "After Removing Bit Shifts:")
         // Run mem2reg once more with the flattened CFG to catch any remaining loads/stores
         .run_pass(Ssa::mem2reg, "After Mem2Reg:")
+        // Run the inlining pass again as certain codegen attributes will now be disabled after flattening,
+        // such as treating functions marked with the `InlineType::Never` as an entry point.
+        .run_pass(Ssa::inline_functions, "After Inlining:")
```
Comment (vezenovm) [inline compiler/noirc_evaluator/src/ssa.rs:66]: Most tests aside a few (specifically `brillig_fns_as_values`, `u128` and `hashmap`) every test passed when the second `inline_functions` was after `mem2reg`

**Comment by jfecher** [comment]
Is this PR ready for review now?

> #[inline(post_if_flatten)] or #[inline(post_flatten)]. It is a bit more verbose but it gets across better what is actually happening and how it provides an optimization when writing circuits. 

I definitely think that we shouldn't call it `#[inline(never)]` if it does actually get inlined eventually.

Also, can you elaborate more on why inlining after flattening fixes #4688? It seems somewhat odd to me - like we're missing a more direct optimization or fix instead.

**Comment by vezenovm** [comment]
> Is this PR ready for review now?

Yeah it is

**Comment by vezenovm** [comment]
> I definitely think that we shouldn't call it `#[inline(never)]` if it does actually get inlined eventually.

We can discuss further during scrum and I can change it as part of this PR or a follow-up. I was leaning towards the more verbose `post_if_flatten`.

> Also, can you elaborate more on why inlining after flattening fixes #4688?

We are getting a big blow-up for poseidon due to how predicates need to be handled when working with large arrays and complex logic. Each call to poseidon is inlined before flattening thus causing everything inside of poseidon also needing to be flattened. In some cases (such as finding a hash) we don't need to worry about handling a predicate appropriately in the failure case and just want the result from the passing case. 

Our black box functions like `pedersen_hash` do not have a blow-up like this as they lay down a single "Call" instruction and do not inlined anything. We then only flatten the result of the hash if it is used outside of the if statement. 

For this program using `pedersen_hash`:
```rust
pub fn pedersen_hash<N>(inputs: [Field; N], hash_index: u32) -> Field {
    dep::std::hash::pedersen_hash_with_separator(inputs, hash_index)
}
fn main(
    to_hash: [[Field; HASH_LENGTH]; NUM_HASHES],
    enable: [bool; NUM_HASHES]
) -> pub [Field; NUM_HASHES + 1] {
    let mut result = [0; NUM_HASHES + 1];
    for i in 0..NUM_HASHES {
        let enable = enable[i];
        let to_hash = to_hash[i];
        if enable {
            result[i] = pedersen_hash(to_hash, 0);
            // result[i] = poseidon_hash(to_hash);
        }
    }
    result
}
``` 
We have the following SSA:
```
After Array Set Optimizations:
acir(inline) fn main f0 {
  b0(v0: [[Field; 10]; 2], v1: [u1; 2]):
    v55 = array_get v1, index u64 0
    v56 = array_get v0, index u64 0
    v57 = call pedersen_hash(v56, u32 0)
    enable_side_effects u1 1
    v59 = cast v55 as Field
    v61 = mul v59, v57
    v62 = array_get v1, index u64 1
    v63 = array_get v0, index u64 1
    v64 = call pedersen_hash(v63, u32 0)
    v65 = not v62
    enable_side_effects u1 1
    v66 = cast v62 as Field
    v67 = cast v65 as Field
    v68 = mul v66, v61
    v69 = mul v67, v61
    v70 = add v68, v69
    v71 = mul v66, v64
    return [v70, v71, Field 0]
}
```
Replacing pedersen with `poseidon_hash` without an inline attribute leads to a massive SSA (where I will not post it as it is up to 12k values). Enabling the developer to inline post flattening allows them to write their own black box function call in a way, and escape some of the compiler's codegen. 

> It seems somewhat odd to me - like we're missing a more direct optimization or fix instead.

I definitely think it would still be good to optimize array merges either way. However, this felt like it could be its own issue and I'm not sure if it would still provide as meaningful optimization that this inlining attribute provides. It actually felt kind of natural to me that Noir would start to adopt codegen attributes that can provide optimizations to the user.

**Comment by jfecher** [comment]
> Enabling the developer to inline post flattening allows them to write their own black box function call in a way, and escape some of the compiler's codegen.

The tradeoff of this would be if there are failing constraints or similar inside the function call then they'd still fail even if the call was wrapped by an `if false { ... }` though, right? I think we should discuss this more if so because that sounds very confusing to me if that is the case.

> I definitely think it would still be good to optimize array merges either way. However, this felt like it could be its own issue and I'm not sure if it would still provide as meaningful optimization that this inlining attribute provides. It actually felt kind of natural to me that Noir would start to adopt codegen attributes that can provide optimizations to the user.

Right, that comment wasn't in reference to array merges which barely affected the original issue constraint count wise. I also don't disagree with adding more compilation/optimization annotations. I was just thinking that this particular solution seemed indirect and I was wondering if there was a more direct way to optimize out predicates we're missing in general. Assuming my `if false { ... }` question above is wrong, one issue with this approach would be that it requires users to know about this somewhat niche problem and apply an annotation fix manually. That isn't the end of the world but it did make me wonder if there was an alternative.

**Comment by jfecher** [comment]
> The tradeoff of this would be if there are failing constraints or similar inside the function call then they'd still fail even if the call was wrapped by an if false { ... } though, right? I think we should discuss this more if so because that sounds very confusing to me if that is the case.

I wonder if we can have an optimization or change semantics in a way that we can only apply predicates to results in general instead of every intermediate step. Maybe we'd need to check for side-effect boundaries and delay it until then? Or group up side effects to perform all at once later somehow? Not sure.

**Comment by jfecher** [comment]
Looks like:
```rs
fn main(x: bool) {
    if x {
        foo();
    }
}

#[inline(never)]
fn foo() {
    assert(false);
}
```
with x = false fails

**Comment by jfecher** [comment]
Discussed in standup:

- Name change: I think `no_predicates` or similar would be good since it is more direct in the reason someone may apply the annotation instead of just referencing how we implement that behavior via inlining later on.
- it seems the `if x { foo() }` example may only be failing because of a bug in the remove enable side effects pass. We can fix that separately.
- I was hesitant at first but I suppose it's fine in general to add opt-in optimization annotations that can potentially break programs if used incorrectly. The opt-in and (eventually) documentation around them will be key here.

**Comment by vezenovm** [comment]
> * Name change: I think `no_predicates` or similar would be good since it is more direct in the reason someone may apply the annotation instead of just referencing how we implement that behavior via inlining later on.

I ended up going with `no_predicates`. It most clearly conveys what is happening to the user and the unsafe nature of using the attribute.

File: compiler/noirc_evaluator/src/ssa/ssa_gen/program.rs [hunk 2]
```diff
@@ -17,6 +17,10 @@ pub(crate) struct Ssa {
     /// This mapping is necessary to use the correct function pointer for an ACIR call,
     /// as the final program artifact will be a list of only entry point functions.
     pub(crate) entry_point_to_generated_index: BTreeMap<FunctionId, u32>,
+    /// Flag that determines when to inline functions marked to not use predicates.
+    /// This is currently defaulted to `false` and should be set to true by after completion
+    /// of the flattening pass which sets side effects.
+    pub(crate) past_flattening_pass: bool,
```
Comment (jfecher) [inline compiler/noirc_evaluator/src/ssa/ssa_gen/program.rs:23]:
```suggestion
```
I think it is better to have an alternate entry point into the inlining pass (e.g. `inline_no_predicates` or similar) than to keep track of an extra bit of internal state which is only relevant for enabling one pass.

**Review by jfecher — CHANGES_REQUESTED** [review]

File: compiler/noirc_evaluator/src/ssa/ssa_gen/program.rs (hunk 2 above)
Comment (vezenovm) [reply compiler/noirc_evaluator/src/ssa/ssa_gen/program.rs]: I switched it over

File: compiler/noirc_evaluator/src/ssa/opt/inlining.rs [hunk 3]
```diff
@@ -36,14 +36,28 @@ impl Ssa {
...
+    // Run the inlining pass where functions marked with `InlineType::NoPredicates` as not entry points
+    pub(crate) fn inline_functions_with_no_predicates(self) -> Ssa {
+        Self::inline_functions_inner(self, false)
+    }
 
+    fn inline_functions_inner(mut self, no_predicates_is_entry_point: bool) -> Ssa {
+        self.functions = btree_map(
+            get_entry_point_functions(&self, no_predicates_is_entry_point),
+            |entry_point| {
+                let new_function = InlineContext::new(&self, entry_point, false).inline_all(&self);
```
Comment (jfecher) [inline compiler/noirc_evaluator/src/ssa/opt/inlining.rs:57]: `false -> no_predicates_is_entry_point`? Or is it unneeded?

Comment (vezenovm) [reply compiler/noirc_evaluator/src/ssa/opt/inlining.rs:57]: no this is a bug I just caught this as well

**Review by jfecher — APPROVED** [review]
`no_predicates_is_entry_point` is a bit of an awkward name but I've no better suggestions at this point and it's not worth blocking over

 :+1:

**Comment by vezenovm** [comment]
> `no_predicates_is_entry_point` is a bit of an awkward name

Yeah I thought the same but was unsure about a better one and didn't feel the need to make some context struct for one flag.

**Pull request merged**